pkg/sentry/kernel/msgqueue/msgqueue.go

```diff
@@ -186,6 +216,240 @@ func (r *Registry) Remove(id ipc.ID, creds *auth.Credentials) error {
 	return nil
 }
 
+// FindByID returns the queue with the specified ID and an error if the ID
+// doesn't exist.
+func (r *Registry) FindByID(id ipc.ID) (*Queue, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	mech := r.reg.FindByID(id)
+	if mech == nil {
+		return nil, syserror.EINVAL
+	}
+	return mech.(*Queue), nil
+}
+
+// Send appends a message to the message queue, and returns an error if sending
+// fails. See msgsnd(2).
+func (q *Queue) Send(ctx context.Context, m Message, b Blocker, wait bool, pid int32) error {
+	creds := auth.CredentialsFromContext(ctx)
+	if err := q.append(ctx, m, creds, pid); err != syserror.ErrWouldBlock {
+		return err
+	}
+
+	if !wait {
+		return syserror.EAGAIN
+	}
+
+	e, ch := waiter.NewChannelEntry(nil)
+	q.senders.EventRegister(&e, waiter.EventOut)
+
+	err := q.append(ctx, m, creds, pid)
+	for ; err == syserror.ErrWouldBlock; err = q.append(ctx, m, creds, pid) {
+		b.Block(ch)
+	}
+	q.senders.EventUnregister(&e)
+	return err
+}
+
+// append appends a message to the queue's message list and notifies waiting
+// receivers that a message has been inserted. It returns an error if adding
+// the message would cause the queue to exceed its maximum capacity, which can
+// be used as a signal to block the task. Other errors should be returned as is.
+func (q *Queue) append(ctx context.Context, m Message, creds *auth.Credentials, pid int32) error {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// Queue was removed while the process was waiting.
+	if q.dead {
+		return syserror.EIDRM
+	}
+
+	if m.mType <= 0 {
+		return syserror.EINVAL
+	}
+
+	if !q.obj.CheckPermissions(creds, fs.PermMask{Write: true}) {
+		// The calling process does not have write permission on the message
+		// queue, and does not have the CAP_IPC_OWNER capability in the user
+		// namespace that governs its IPC namespace.
+		return syserror.EACCES
+	}
+
+	// Check if there's space for a message.
+	if m.mSize+q.byteCount > q.maxBytes || q.messageCount+1 > q.maxBytes {
```

Comment:
Something seems wrong here, does the second clause needs to compare against a max message limit?

Reply:
Linux uses the maximum queue size as a limit for message count, this is done to prevent unlimited zero-size messages. Here's a snippet from the man pages
```
A message queue is considered to be full if either of the following
conditions is true:

 • Adding a new message to the queue would cause the total number
   of bytes in the queue to exceed the queue's maximum size (the
   msg_qbytes field).

 • Adding another message to the queue would cause the total
   number of messages in the queue to exceed the queue's maximum
   size (the msg_qbytes field).  This check is necessary to
   prevent an unlimited number of zero-length messages being
   placed on the queue.  Although such messages contain no data,
   they nevertheless consume (locked) kernel memory.
```
I added this comment to the code, with a note that `msg_qbytes` is `q.maxBytes` for us.